api/models/schema.py

```diff
@@ -1,109 +1,283 @@
-                    prop['errorMessage'] = {"pattern": "must match format \"url\""}
+    @classmethod
+    def update_url_schema(cls, url_schema: dict[str, Any]) -> dict[str, Any]:
+        """Update the schema for url fields to use a regex for url matching"""
+        if "format" in url_schema and url_schema["format"] == "uri":
+            url_schema.pop("format")
+            url_schema["pattern"] = url_pattern
+            url_schema["errorMessage"] = {"pattern": url_pattern_error}
+        return url_schema
+
+    @classmethod
+    def remove_any_of(cls, schema: dict[str, Any], field: str) -> dict[str, Any]:
+        """Remove the anyOf key from the schema for the given field and set the value of the field to the first
+        item in the anyOf list - the 2nd item in the list is always {"type": "null"} which we don't want
+        in our schema for the UI to work"""
+
+        field_schema = schema["properties"][field]
+        anyOf = field_schema.pop("anyOf")
+        field_schema.update(anyOf[0])
+        return field_schema
+
+    @classmethod
+    def remove_default_null(cls, field_schema: dict[str, Any]) -> dict[str, Any]:
+        """Remove the default value of null from the schema for the given field"""
+        if "default" in field_schema and field_schema["default"] is None:
+            field_schema.pop("default")
+        return field_schema
+
+    @classmethod
+    def remove_default_empty_list(
+        cls, field_schema: dict[str, Any]
+    ) -> dict[str, Any]:
+        """Remove the default value of an empty list from the schema for the given field"""
+        if "default" in field_schema and field_schema["default"] == []:
+            field_schema.pop("default")
+        return field_schema
 
 
 class CreativeWork(SchemaBaseModel):
     type: str = Field(
         alias="@type",
         default="CreativeWork",
         description="Submission type can include various forms of content, such as datasets, "
-                    "software source code, digital documents, etc.",
+        "software source code, digital documents, etc.",
     )
     name: str = Field(description="Submission's name or title", title="Name or title")
 
 
+class _CreativeWorkMixin(CreativeWork):
+    """Mixin class for CreativeWork models to custom generate json schema"""
+
+    @classmethod
+    def __get_pydantic_json_schema__(
+        cls, core_schema: CoreSchema, handler: GetJsonSchemaHandler
+    ) -> JsonSchemaValue:
+        # adjusting the schema to remove the anyOf key for url and description optional fields
+        # we are doing this because the schema based UI is currently not able to handle the anyOf key with
+        # two possible types - one of which is null (due to the field being optional)
+        json_schema = handler(core_schema)
+        json_schema = handler.resolve_ref_schema(json_schema)
+        for field in ("url", "description"):
```

Comment:
These one off json schema modifiers look hairy.  Any way we can consolidate these changes to the base class and make the changes for certain types of attributes?